torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,124 @@ void removeMaxPoolUnusedOutput(Block* b) {
+// of the input and different attributes there will be different subgraphs of
+// LogSoftmax and NegativeLogLikelihoodLoss.
+static void fuseLogSoftmaxNllLoss(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      fuseLogSoftmaxNllLoss(child_block);
+    }
+    if (it->kind() == onnx::NegativeLogLikelihoodLoss) {
+      auto prev = it->input(0)->node();
+      Node* origNllLossNode = *it;
+      Node* origLogSoftmaxNode;
+      if (prev->kind() == onnx::LogSoftmax) {
+        // if the input is 2D
+        // graph(%input : Float(3, 5),
+        // %target : Long(3)):
+        // %4 : Float(3, 5) = onnx::LogSoftmaxaxis=1
+        // %8 : Float(3) = onnx::NegativeLogLikelihoodLoss[reduction="none"]
+        // return (%8)
+        origLogSoftmaxNode = it->input(0)->node();
+      } else if (
+          prev->kind() == onnx::Transpose &&
+          prev->input(0)->node()->kind() == onnx::LogSoftmax) {
+        // if the input is 4D
+        // graph(%input : Float(3, 5, 2, 7),
+        // %target : Long(3, 2, 7)):
+        // %4 : Tensor = onnx::Transpose[perm=[0, 3, 2, 1]] (%input)
+        // %5 : Tensor = onnx::LogSoftmax[axis=3] (%4)
+        // %6 : Float(3, 5, 2, 7) = onnx::Transpose[perm=[0, 3, 2, 1]] (%5)
+        // %10 : Float(3, 2, 7) =
+        // onnx::NegativeLogLikelihoodLoss[reduction="none"](%6, %target) return
+        // (%10)
+        origLogSoftmaxNode = prev->input(0)->node();
+        auto transpose = origLogSoftmaxNode->input(0)->node();
+        origLogSoftmaxNode->replaceInput(0, transpose->inputs().at(0));
+      } else if (
+          prev->kind() == onnx::Reshape &&
+          prev->input(0)->node()->kind() == onnx::Transpose &&
+          prev->input(0)->node()->input(0)->node()->kind() ==
+              onnx::LogSoftmax) {
+        // if the input is 3D or > 4D
+        // graph(%input : Float(3, 5, 2),
+        // %target.1 : Long(3, 2)):
+        // %4 : Tensor = onnx::Transpose[perm=[0, 2, 1]] (%input)
+        // %5 : Tensor = onnx::LogSoftmax[axis=2] (%4)
+        // %6 : Float(3, 5, 2) = onnx::Transpose[perm=[0, 2, 1]] (%5)
+        // %8 : Tensor = onnx::Shape(%6)
+        // %10 : Tensor = onnx::Constantvalue={0}
+        // %11 : Long() = onnx::Gather[axis=0] (%8, %10)
+        // %13 : Tensor = onnx::Shape(%6)
+        // %15 Tensor = onnx::Constantvalue={1}
+        // %16 : Long() = onnx::Gather[axis=0] (%13, %15)
+        // ...
+        // %22 : Float(3, 5, 1, 2) = onnx::Reshape(%6, %21)
+        // ...
+        // %26 : Long(3, 1, 2) = onnx::Reshape(%target.1, %25)
+        // %30 : Float() = onnx::NegativeLogLikelihoodLoss[reduction="sum"](%22,
+        // %26) return (%30)
+        TORCH_INTERNAL_ASSERT(
+            prev->input(1)->node()->input(0)->node()->kind() == onnx::Gather);
+        TORCH_INTERNAL_ASSERT(
+            prev->input(1)->node()->input(1)->node()->kind() == onnx::Gather);
```

Comment:
@KsenijaS My PR https://github.com/pytorch/pytorch/pull/54569 has triggered these two ASSERT failures. May I know if these two ASSERT is necessary? Just checked the below logic, but don't see any places uses `Gather` property. Thanks